File: llm_generated_code/Gemini/CPP/Level2/Q33.cpp
#include <iostream>
#include <vector>
#include <climits>

using namespace std;

int maxProfit(const vector<int>& prices) {
    int n = prices.size();
    if (n <= 1) {
        return 0;
    }
    vector<int> profits(n);
    int max_price = prices[n - 1];
    for (int i = n - 2; i >= 0; i--) {
        max_price = max(max_price, prices[i]);
        profits[i] = max(profits[i + 1], max_price - prices[i]);
    }
    int min_price = prices[0];
    int max_profit = 0;
    for (int i = 1; i < n; i++) {
        max_profit = max(max_profit, profits[i] + max(prices[i] - min_price, 0));
        min_price = min(min_price, prices[i]);
    }

    return max_profit;
}

int main() {
    vector<int> prices = {7, 1, 5, 3, 6, 4}; 

    int max_profit = maxProfit(prices);
    cout << "Maximum profit: " << max_profit << endl;

    return 0;
}

File: llm_generated_code/Gemini/Python/Level2/Q33.py
def max_profit(prices):
    n = len(prices)
    profit = 0

    max_profit_once = 0
    min_price = prices[0]
    for i in range(1, n):
        max_profit_once = max(max_profit_once, prices[i] - min_price)
        min_price = min(min_price, prices[i])

    max_profit_twice = 0
    max_so_far = prices[n - 1]
    for i in range(n - 2, -1, -1):
        max_profit_twice = max(max_profit_twice, max_so_far - prices[i] + max_profit_once)
        max_so_far = max(max_so_far, prices[i])

    return max(max_profit_once, max_profit_twice)

def main():
    n = int(input())
    prices = list(map(int, input().split()))
    print(max_profit(prices))

if __name__ == "__main__":
    main()
